Add SPIR-V NonSemantic.Shader.DebugInfo.200

MrSidims opened the pull request:
It's based on NonSemantic.Shader.DebugInfo.100

New functionality compared to NonSemantic.Shader.DebugInfo.100
New instructions: DebugModule, DebugTypeSubrange,
DebugTypeArrayDynamic and DebugTypeString.

New Debug Operations.

DebugTypeArray no longer has an `Component Count <id>`
operand. It was replaced by a new operand `<id>, <id>, ...
Subranges` to fully represent all characteristics of the
elements of DebugTypeArray.

Added an optional 'Target function' operand to
DebugFunction.

Added a 'SrcLang' table to be used by DebugCompilationUnit
instruction.

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

MrSidims commented:
@baldurk hi, may I ask you to take a look? The patch adds new debug instruction, mostly for Fortran and also extends number of debug operations to match DWARF (though some of them require analog of .debug_address and hence are currently just placeholders).

Initially we called the extension as NonSemantic.Kernel.DebugInfo.100 to outline it as debug information for compute and to put less pressure for other vendors to implement it. If you want - this can be restored, but TBH having a single extension looks better.

MrSidims commented on nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc:
```diff
@@ -111,13 +116,31 @@ This is a non-normative list of changes to the OpenCL.DebugInfo.100 specificatio
 * <<DebugBasicType,*DebugBasicType*>> now takes a *Flags* operand to allow specifying
   *FlagUnknownPhysicalLayout*.
 
+TODO: to remove, kept for a while for review
+Additional list of changes to the NonSemantic.Shader.DebugInfo.100 specification rev 9:
```
Relic of NonSemantic.Kernel.DebugInfo.100 . Kept for now, but should be removed.

baldurk commented:
I don't think it's a good idea to make significant revisions to this extinst set after it has been released. Generally speaking that's avoided in vulkan and SPIR-V extensions to my knowledge and I think the ecosystem impact would be hard to predict.

I'd also point you to the original `OpenCL.DebugInfo.100` extended instruction set, this one is intended for graphics and graphics shaders, so if you're doing OpenCL work that instruction set (or a new one based on it) is likely a better fit. That extension is also managed by Intel I believe so you might be able to discuss it internally too.

bashbaug commented:
We've discussed this internally already and our current thinking is that a) it'd be better to use a non-semantic extended instruction set going forward vs. the original `OpenCL.DebugInfo.100` extended instruction set, and b) it'd be nice to have one unified extended instruction set for debug information vs. two different extended instruction sets that are either potentially different or that we'd need to keep in sync.

We're open to suggestions how best to do this.  Perhaps we should do a `NonSemantic.Shader.DebugInfo.XYZ` version instead (for some values of `XYZ`) so it is clear this is a newer version?

baldurk commented:
In that case yes I think the best thing to do is make a new extended instruction set heavily based on this one, as this one was based on the OpenCL one. Either NonSemantic.Shader.DebugInfo.101 or NonSemantic.Shader.DebugInfo.200, whatever you think is best. Bear in mind though I don't expect the existing users of this set will update to it, so there's still going to be multiple sets in that sense (this version, any new version, and OpenCL.DebugInfo.100).

Commit:
Add SPIR-V NonSemantic Kernel DebugInfo Instructions

It's based on NonSemantic.Shader.DebugInfo.100

New functionality compared to NonSemantic.Shader.DebugInfo.100
New instructions: DebugModule, DebugTypeSubrange,
DebugTypeArrayDynamic and DebugTypeString.

New Debug Operations.

DebugTypeArray no longer has an `Component Count <id>`
operand. It was replaced by a new operand `<id>, <id>, ...
Subranges` to fully represent all characteristics of the
elements of DebugTypeArray.

Added an optional 'Target function' operand to
DebugFunction.

Added a 'SrcLang' table to be used by DebugCompilationUnit
instruction.

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

MrSidims commented:
Thanks! Based on the feedback restored the original name of the instruction set.

Commit:
NonSemantic.Kernel.DebugInfo.100 -> NonSemantic.Shader.DebugInfo.200

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

Commit:
Fix parameters and typos

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

Commit:
Add plus

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

dgkoch commented on nonsemantic/NonSemantic.Shader.DebugInfo.200.asciidoc:
```diff
@@ -0,0 +1,815 @@
+ - David Neto, Google
+ - Neil Henning, Codeplay
+ - Kerch Holt, Nvidia
+ - Jaebaek Seo, Google
```
I would not re-list the contributors to other specs by name. (similar to how we don't re-list the instructions in the .100 specs).

Commit:
Languages reordering + correct contribution list

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

Commit:
Merge remote-tracking branch 'origin/main' into update-nonsemantic-debug

Commit:
Add Fortran 2018 and C17

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

bashbaug commented:
@MrSidims do we have SPIR-V headers and any SPIR-V tools changes done for these updates?  This would be helpful to gain confidence that the "200" version of this extended instruction set layers properly over the "100" version.  Thanks!

MrSidims commented:
Moved to draft for a while to figure out how to express virtuality.

Commit:
Update for the spec

1. added complex type encoding
2. added bitfield flag
3. fixed debug operations
4. added inline namespace

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

Commit:
Add checksum and column to inlinedat

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

Commit:
Clarify count and upper bound of TypeSubrange

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

Commit:
Add 'Memory Size' field to DebugTypeVector

Signed-off-by: Lu, John <john.lu@intel.com>

Commit:
Revert "Add 'Memory Size' field to DebugTypeVector"

This reverts commit 4cdd415632540b08085c895c2a2a9a383688f9ef.

Calculation of memory size is done with (element size)*bit_ceil(# elements). 'Memory Size' field is not needed.

spencer-lunarg commented:
Now that we have a published `NonSemantic.Shader.DebugInfo.101`, can we go that route and add as much "additive" stuff here and make a `NonSemantic.Shader.DebugInfo.102` instead?

It seems the only thing you can't "add" is `DebugTypeArray` but you could honestly make a `DebugTypeArray2` at this point

It is going to MUCH easier to have tools "ignore" unknown things than having to manage 2 complete, 90% the same DebugInfo instruction sets

MrSidims commented:
> can we go that route and add as much "additive" stuff here and make a NonSemantic.Shader.DebugInfo.102 instead

Yes, it's on my TODO list to start porting few things to the rolling version of debug info. Though Fortran-specific things (like TypeArray) should probably be ported by somebody else who are interested in offloading of Fortran through SPIR-V :) . I'd keep this PR open for a while, also I don't expect .200 extension to be completely abandon for few reasons, which @vmaksimo might (or might not) comment on.
